File: packages/controlplane/src/domain/entities/semantic_registry.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct SemanticDefinition {
    pub id: Uuid,
    pub namespace: String,
    pub name: String,
    pub object_type: String,
    pub version: i32,
    pub status: String,
    pub owner_principal: String,
    pub description: String,
    pub spec: serde_json::Value,
    pub time_semantics: Option<serde_json::Value>,
    pub supersedes_definition_id: Option<Uuid>,
    pub deprecation_deadline: Option<DateTime<Utc>>,
    pub created_by: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct SemanticDependency {
    pub id: Uuid,
    pub source_definition_id: Uuid,
    pub target_definition_id: Uuid,
    pub dependency_type: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct SemanticPhysicalDependency {
    pub id: Uuid,
    pub semantic_definition_id: Uuid,
    pub catalog: String,
    pub schema_name: String,
    pub object_name: String,
    pub object_type: String,
    pub contract_version: Option<i32>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct SemanticLifecycleEvent {
    pub id: Uuid,
    pub definition_id: Uuid,
    pub previous_status: Option<String>,
    pub new_status: String,
    pub principal: String,
    pub reason: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SemanticDependencyRef {
    pub namespace: String,
    pub name: String,
    pub version: i32,
    #[serde(default = "default_dependency_type")]
    pub dependency_type: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SemanticPhysicalDependencyInput {
    pub catalog: String,
    pub schema_name: String,
    pub object_name: String,
    #[serde(default = "default_physical_object_type")]
    pub object_type: String,
    #[serde(default)]
    pub contract_version: Option<i32>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateSemanticDefinitionRequest {
    pub namespace: String,
    pub name: String,
    #[serde(default = "default_object_type")]
    pub object_type: String,
    pub version: i32,
    pub owner_principal: String,
    #[serde(default)]
    pub description: String,
    pub spec: serde_json::Value,
    #[serde(default)]
    pub time_semantics: Option<serde_json::Value>,
    #[serde(default)]
    pub supersedes_version: Option<i32>,
    #[serde(default)]
    pub dependencies: Vec<SemanticDependencyRef>,
    #[serde(default)]
    pub physical_dependencies: Vec<SemanticPhysicalDependencyInput>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct TransitionSemanticStatusRequest {
    pub status: String,
    #[serde(default)]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SemanticDefinitionsQuery {
    #[serde(default)]
    pub search: Option<String>,
    #[serde(default)]
    pub namespace: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub object_type: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SemanticGraphQuery {
    #[serde(default)]
    pub direction: Option<String>,
    #[serde(default)]
    pub depth: Option<usize>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SemanticCompareQuery {
    pub from: i32,
    pub to: i32,
}

#[derive(Debug, Clone, Serialize)]
pub struct SemanticDefinitionSummary {
    pub namespace: String,
    pub name: String,
    pub object_type: String,
    pub owner_principal: String,
    pub description: String,
    pub active_version: Option<i32>,
    pub latest_version: i32,
    pub latest_status: String,
    pub version_count: usize,
    pub semantic_dependency_count: usize,
    pub direct_dependent_count: usize,
    pub physical_dependency_count: usize,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize)]
pub struct SemanticDefinitionDetail {
    pub definition: SemanticDefinition,
    pub dependencies: Vec<SemanticDefinition>,
    pub dependency_edges: Vec<SemanticDependency>,
    pub dependents: Vec<SemanticDefinition>,
    pub physical_dependencies: Vec<SemanticPhysicalDependency>,
    pub lifecycle_history: Vec<SemanticLifecycleEvent>,
}

#[derive(Debug, Clone, Serialize)]
pub struct SemanticGraphResponse {
    pub root: SemanticDefinition,
    pub direction: String,
    pub depth: usize,
    pub nodes: Vec<SemanticDefinition>,
    pub edges: Vec<SemanticDependency>,
}

#[derive(Debug, Clone, Serialize)]
pub struct SemanticCompareResponse {
    pub from: SemanticDefinitionDetail,
    pub to: SemanticDefinitionDetail,
    pub changes: serde_json::Value,
}

fn default_dependency_type() -> String {
    "semantic".to_string()
}

fn default_physical_object_type() -> String {
    "table".to_string()
}

fn default_object_type() -> String {
    "metric".to_string()
}

File: packages/webui/services/semantic-registry.ts
import { apiFetch } from "@/lib/api-client"

export type SemanticStatus =
  | "draft"
  | "validated"
  | "candidate"
  | "certified"
  | "active"
  | "deprecated"
  | "retired"

export interface SemanticDefinition {
  created_at: string
  created_by: string
  deprecation_deadline: string | null
  description: string
  id: string
  name: string
  namespace: string
  object_type: "metric"
  owner_principal: string
  spec: unknown
  status: SemanticStatus
  supersedes_definition_id: string | null
  time_semantics: unknown | null
  updated_at: string
  version: number
}

export interface SemanticDependency {
  created_at: string
  dependency_type: string
  id: string
  source_definition_id: string
  target_definition_id: string
}

export interface SemanticPhysicalDependency {
  catalog: string
  contract_version: number | null
  created_at: string
  id: string
  object_name: string
  object_type: string
  schema_name: string
  semantic_definition_id: string
}

export interface SemanticLifecycleEvent {
  created_at: string
  definition_id: string
  id: string
  new_status: SemanticStatus
  previous_status: SemanticStatus | null
  principal: string
  reason: string | null
}

export interface SemanticDefinitionSummary {
  active_version: number | null
  description: string
  direct_dependent_count: number
  latest_status: SemanticStatus
  latest_version: number
  name: string
  namespace: string
  object_type: "metric"
  owner_principal: string
  physical_dependency_count: number
  semantic_dependency_count: number
  updated_at: string
  version_count: number
}

export interface SemanticDefinitionDetail {
  definition: SemanticDefinition
  dependencies: SemanticDefinition[]
  dependency_edges: SemanticDependency[]
  dependents: SemanticDefinition[]
  lifecycle_history: SemanticLifecycleEvent[]
  physical_dependencies: SemanticPhysicalDependency[]
}

export interface CreateSemanticDefinitionBody {
  dependencies?: Array<{
    namespace: string
    name: string
    version: number
    dependency_type?: string
  }>
  description: string
  name: string
  namespace: string
  object_type?: "metric"
  owner_principal: string
  physical_dependencies?: Array<{
    catalog: string
    schema_name: string
    object_name: string
    object_type?: string
    contract_version?: number | null
  }>
  spec: unknown
  supersedes_version?: number | null
  time_semantics?: unknown | null
  version: number
}

export interface SemanticCompareResponse {
  changes: Record<string, boolean>
  from: SemanticDefinitionDetail
  to: SemanticDefinitionDetail
}

export async function listSemanticDefinitions(params?: {
  search?: string
  namespace?: string
  status?: string
}): Promise<SemanticDefinitionSummary[]> {
  const url = new URL(
    "/api/semantic-registry/definitions",
    window.location.origin
  )
  if (params?.search) {
    url.searchParams.set("search", params.search)
  }
  if (params?.namespace) {
    url.searchParams.set("namespace", params.namespace)
  }
  if (params?.status && params.status !== "all") {
    url.searchParams.set("status", params.status)
  }
  const res = await apiFetch(url.toString())
  if (!res.ok) {
    throw await responseError(res)
  }
  const body = await res.json()
  return body.definitions
}

export async function createSemanticDefinition(
  body: CreateSemanticDefinitionBody
): Promise<SemanticDefinitionDetail> {
  const res = await apiFetch("/api/semantic-registry/definitions", {
    body: JSON.stringify(body),
    headers: { "Content-Type": "application/json" },
    method: "POST",
  })
  if (!res.ok) {
    throw await responseError(res)
  }
  return res.json()
}

export async function createSemanticVersion(
  namespace: string,
  name: string,
  body: CreateSemanticDefinitionBody
): Promise<SemanticDefinitionDetail> {
  const res = await apiFetch(
    `/api/semantic-registry/definitions/${encodeURIComponent(namespace)}/${encodeURIComponent(name)}/versions`,
    {
      body: JSON.stringify(body),
      headers: { "Content-Type": "application/json" },
      method: "POST",
    }
  )
  if (!res.ok) {
    throw await responseError(res)
  }
  return res.json()
}

export async function listSemanticVersions(
  namespace: string,
  name: string
): Promise<SemanticDefinition[]> {
  const res = await apiFetch(
    `/api/semantic-registry/definitions/${encodeURIComponent(namespace)}/${encodeURIComponent(name)}`
  )
  if (!res.ok) {
    throw await responseError(res)
  }
  const body = await res.json()
  return body.versions
}

export async function getSemanticDefinition(
  namespace: string,
  name: string,
  version: number
): Promise<SemanticDefinitionDetail> {
  const res = await apiFetch(
    `/api/semantic-registry/definitions/${encodeURIComponent(namespace)}/${encodeURIComponent(name)}/versions/${version}`
  )
  if (!res.ok) {
    throw await responseError(res)
  }
  return res.json()
}

export async function transitionSemanticStatus(
  namespace: string,
  name: string,
  version: number,
  status: SemanticStatus,
  reason?: string
): Promise<SemanticDefinitionDetail> {
  const res = await apiFetch(
    `/api/semantic-registry/definitions/${encodeURIComponent(namespace)}/${encodeURIComponent(name)}/versions/${version}/status`,
    {
      body: JSON.stringify({ reason: reason ?? null, status }),
      headers: { "Content-Type": "application/json" },
      method: "PATCH",
    }
  )
  if (!res.ok) {
    throw await responseError(res)
  }
  return res.json()
}

export async function compareSemanticVersions(
  namespace: string,
  name: string,
  from: number,
  to: number
): Promise<SemanticCompareResponse> {
  const url = new URL(
    `/api/semantic-registry/definitions/${encodeURIComponent(namespace)}/${encodeURIComponent(name)}/compare`,
    window.location.origin
  )
  url.searchParams.set("from", String(from))
  url.searchParams.set("to", String(to))
  const res = await apiFetch(url.toString())
  if (!res.ok) {
    throw await responseError(res)
  }
  return res.json()
}

async function responseError(res: Response) {
  const err = await res.json().catch(() => ({}))
  return new Error((err as { error?: string }).error ?? `HTTP ${res.status}`)
}
